### PR TITLE
adding ranged number type, for use in stats

### PR DESCRIPTION
**Pull request by gravestench**
unit tests included

### CODE REVIEW

File: d2common/d2math/ranged_number.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// Mul multiplies this ranged number by the given ranged number, returning this one
+func (rn *RangedNumber) Mul(other *RangedNumber) *RangedNumber {
+	return rn.Set(rn.min * other.min, rn.max * other.max)
+}
+
+// Div divides this ranged number by the given ranged number, returning this one
+func (rn *RangedNumber) Div(other *RangedNumber) *RangedNumber {
+	return rn.Set(rn.min / other.min, rn.max / other.max)
+}
+
```
Comment (malavv) [inline d2common/d2math/ranged_number.go:93]: I'm sure some wouldn't be happy to add something we have no use for yet, but I often see these kind of type with methods : `IsIn(n int)`, `overlaps(rng RangedNumber)`, `encloses(rng RangedNumber)`. Potentially for another PR down the line.

File: d2common/d2math/ranged_number.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+// RangedNumber is a number with a min and max range
+type RangedNumber struct {
+	min int
+	max int
+}
+
+// Min returns the min value, swapping min/max if not ordered
+func (rn *RangedNumber) Min() int {
+	if rn.min > rn.max {
```
[the remark this answers is not in the document]
Comment (gravestench) [reply d2common/d2math/ranged_number.go:13]: I'm not sure, it was worrying me that they could be out of order so I added this, but you're probably right.

File: d2common/d2math/ranged_number.go (hunk 1 above)
Comment (gravestench) [reply d2common/d2math/ranged_number.go:93]: This was originally on my feature branch for stats, but as I finished it I figured I would push now (to keep the upcoming stat PR small)

**Comment by gravestench** [comment]
**Should these be immutable?** 

All of the methods currently alter the struct. I have a hunch that this is probably a bad thing, and that it would be easier to use if the methods returned altered clones.

**Comment by malavv** [comment]
> **Should these be immutable?**
> 
> All of the methods currently alter the struct. I have a hunch that this is probably a bad thing, and that it would be easier to use if the methods returned altered clones.

Good point, I can think of good points for both so I don't have any specific preference. What you've done sounds good, and might be more efficient down the line memory wise.

**Pull request merged**